adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/MqttRequestTenantAndAuthIdExtractor.java

```diff
@@ -0,0 +1,96 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.mqtt;
+
+import java.net.HttpURLConnection;
+
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.client.TenantClientFactory;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.service.BasicTenantAndAuthIdExtractor;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.eclipse.hono.util.TenantObjectWithAuthId;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.mqtt.MqttEndpoint;
+
+/**
+ * Extracts the tenant and auth-id out of an MQTT request.
+ */
+public class MqttRequestTenantAndAuthIdExtractor extends BasicTenantAndAuthIdExtractor {
+
+    /**
+     * Creates a new MqttRequestTenantAndAuthIdExtractor.
+     *
+     * @param config The configuration.
+     * @param tenantClientFactory The factory to use for creating a Tenant service client.
+     * @throws NullPointerException if either of the parameters is {@code null}.
+     */
+    public MqttRequestTenantAndAuthIdExtractor(final ProtocolAdapterProperties config,
+            final TenantClientFactory tenantClientFactory) {
+        super(config, tenantClientFactory);
+    }
+
+    /**
+     * Extracts the tenant and auth-id out of either the client certificate or the username/password
+     * credentials provided in the MQTT CONNECT packet.
+     *
+     * @param endpoint The MQTT endpoint to extract the data out of.
+     * @param spanContext The OpenTracing context to use for tracking the operation (may be {@code null}).
+     * @return A future indicating the outcome of the operation.
+     * @throws NullPointerException if endpoint is {@code null}.
+     */
+    public Future<TenantObjectWithAuthId> getFromMqttEndpoint(final MqttEndpoint endpoint, final SpanContext spanContext) {
+
+        return getFromClientCertificate(endpoint.sslSession(), null, spanContext)
+                .recover(thr -> getFromAuthHeader(endpoint, spanContext));
+    }
+
+    private Future<TenantObjectWithAuthId> getFromAuthHeader(final MqttEndpoint endpoint, final SpanContext spanContext) {
```

Comment:
getFromConnectPacket?

Reply:
I've used getFromMqttAuth now because the method may also get called in the context of a PUBLISH packet.